### PR TITLE
Start stop apps

### PR DESCRIPTION
**Pull request by FrankBakkerNl**
## Proposed change

A lot of refactoring to make non-base class apps have mostly the same behaviour as NetDaemonAppBase apps

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code compiles without warnings (code quality chek)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/net-daemon/docs

### CODE REVIEW

File: tests/NetDaemon.Daemon.Tests/FakesTests/FakeTests.cs [hunk 1]
```diff
@@ -4,6 +4,7 @@
 using System.Globalization;
 using System.Linq;
 using System.Reactive.Linq;
+using System.Security.Cryptography.Xml;
```
Comment (helto4real) [inline tests/NetDaemon.Daemon.Tests/FakesTests/FakeTests.cs:7]: Why is that needed. Can´t see that :)

File: tests/NetDaemon.Daemon.Tests/DaemonRunner/App/DaemonAppTests.cs [hunk 2]
```diff
@@ -34,7 +36,8 @@ public class AppTests
             AppSource = appSource
         });
 
-        private static ServiceProvider ServiceProvider
+
+        private IServiceCollection ServiceCollection
```
Comment (helto4real) [inline tests/NetDaemon.Daemon.Tests/DaemonRunner/App/DaemonAppTests.cs:40]: I am not sure about logic in properties think that is a anti-pattern. Should be a function imo

File: src/App/NetDaemon.App/DependencyInjectionSetup.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+using Microsoft.Extensions.DependencyInjection;
+using NetDaemon.Common;
+using NetDaemon.Daemon.Services;
+
+namespace NetDaemon
+{
+    /// <summary>
+    /// Provides setup for services from NetDaemon.App
+    /// </summary>
+    public static class DependencyInjectionSetup
```
Comment (helto4real) [inline src/App/NetDaemon.App/DependencyInjectionSetup.cs:11]: Maybe call it "DependencyInjectionSetupManager" or smth?

File: src/App/NetDaemon.App/Common/ApplicationContexts/ApplicationContext.cs [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+using System.Diagnostics.CodeAnalysis;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace NetDaemon.Common
+{
+    /// <summary>
+    /// Context for NetDaemon application
+    /// </summary>
+    public abstract class ApplicationContext : IAsyncDisposable, IApplicationMetadata
```
Comment (helto4real) [inline src/App/NetDaemon.App/Common/ApplicationContexts/ApplicationContext.cs:12]: I really like the idea of ApplicationContext!

**Review by helto4real — COMMENTED** [review]
Great job @FrankBakkerNl ! Some minor comments. Think we are good with this!

File: tests/NetDaemon.Daemon.Tests/FakesTests/FakeTests.cs (hunk 1 above)
Comment (FrankBakkerNl) [reply tests/NetDaemon.Daemon.Tests/FakesTests/FakeTests.cs:7]: I have absolutely no idea, must have been added by a refactoring

File: src/App/NetDaemon.App/Common/ApplicationContexts/ApplicationContext.cs [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+            else
+            {
+                return new NonBaseApplicationContext(applicationType, id, serviceProvider);
+            }
+        }
+
+        /// <summary>
+        /// Intended for internal use by fakes only
+        /// </summary>
+        public static ApplicationContext CreateFromAppInstance(INetDaemonAppBase applicationInstance, IServiceProvider serviceProvider)
```
[the remark this answers is not in the document]
Comment (FrankBakkerNl) [reply src/App/NetDaemon.App/Common/ApplicationContexts/ApplicationContext.cs:40]: Maybe that's better indeed. I really wanted to keep ApplictaionCOnetxt as an internal class but that was not easy because of the dependencies with other classes in different projects. Maybe later on this can be improved so it can become internal for now making the name explicit may be best

File: src/App/NetDaemon.App/DependencyInjectionSetup.cs (hunk 3 above)
Comment (FrankBakkerNl) [reply src/App/NetDaemon.App/DependencyInjectionSetup.cs:11]: I added similar `DependencyInjectionSetup` classes to other projects previously. 'Manager' seems a bit overkill to me here. A quick internet search shows ppl sometimes call this ServicesConfiguration or IServcieCollectionExtension. Cannot really fins a 'definitive' best name. Having a file called DependencyInjectionSetup.cs in the root of each project does make it easy to find. Because they are extension methods the class name usually does not end up in usercode so It does not really matter a lot I think

Comment (helto4real) [reply src/App/NetDaemon.App/DependencyInjectionSetup.cs:11]: I see, That is ok.

**Review by helto4real — APPROVED** [review]
